cmd/oras/push.go

```diff
@@ -120,94 +120,178 @@ func runPush(opts pushOptions) error {
+		committed.Store(desc.Digest.String(), desc.Annotations[ocispec.AnnotationTitle])
+		if err := display.PrintSuccessorStatus(ctx, desc, "Skipped  ", store, committed, verbose); err != nil {
+			return err
 		}
-		desc.Annotations = packOpts.ConfigAnnotations
-		packOpts.ConfigDescriptor = &desc
-		packOpts.PackImageManifest = true
+		return display.PrintStatus(desc, "Uploaded ", verbose)
 	}
-	descs, err := loadFiles(ctx, store, annotations, opts.FileRefs, opts.Verbose)
+}
+
+type packFunc func() (ocispec.Descriptor, error)
+type copyFunc func(desc ocispec.Descriptor) error
+
+func pushArtifact(dst *remote.Repository, pack packFunc, packOpts *oras.PackOptions, copy copyFunc, copyOpts *oras.CopyGraphOptions, verbose bool) (ocispec.Descriptor, error) {
+	root, err := pack()
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
 
-	// pack artifact
-	manifestDesc, err := oras.Pack(ctx, store, opts.artifactType, descs, packOpts)
+	copyRootAttempted := false
+	preCopy := copyOpts.PreCopy
+	copyOpts.PreCopy = func(ctx context.Context, desc ocispec.Descriptor) error {
+		if content.Equal(root, desc) {
+			// copyRootAttempted helps track whether the returned error is
+			// generated from copying root.
+			copyRootAttempted = true
+		}
+		if preCopy != nil {
+			return preCopy(ctx, desc)
+		}
+		return nil
+	}
+
+	// push
+	if err = copy(root); err == nil {
+		return root, nil
+	}
+
+	if !copyRootAttempted || !isArtifactUnsupported(err) {
+		return ocispec.Descriptor{}, err
+	}
+
+	if err := display.PrintStatus(root, "Fallback ", verbose); err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	dst.SetReferrersCapability(false)
+	packOpts.PackImageManifest = true
+	root, err = pack()
 	if err != nil {
 		return ocispec.Descriptor{}, err
 	}
 
-	if err = store.Tag(ctx, manifestDesc, tagStaged); err != nil {
+	copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if content.Equal(node, root) {
+			// skip non-config
+			content, err := content.FetchAll(ctx, fetcher, root)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(content, &manifest); err != nil {
+				return nil, err
+			}
+			return []ocispec.Descriptor{manifest.Config}, nil
+		}
+
+		// config has no successors
+		return nil, nil
+	}
+	if err = copy(root); err != nil {
 		return ocispec.Descriptor{}, err
 	}
-	return manifestDesc, nil
+	return root, nil
+}
+
+func isArtifactUnsupported(err error) bool {
+	var errResp *errcode.ErrorResponse
+	if !errors.As(err, &errResp) || errResp.StatusCode != http.StatusBadRequest {
+		return false
+	}
+
+	var errCode errcode.Error
+	if !errors.As(errResp, &errCode) {
+		return false
+	}
+
+	// As of November 2022, ECR is known to return UNSUPPORTED error when
+	// putting an OCI artifact manifest.
+	switch errCode.Code {
+	case errcode.ErrorCodeManifestInvalid, errcode.ErrorCodeUnsupported:
```

Comment:
Probably a non-issue, just curious how do we deal with the casing (upper-case/lower-case) here? (e.g., `UNSUPPORTED` vs `Unsupported`)

```
Error: PUT "https://080565210187.dkr.ecr.us-east-1.amazonaws.com/v2/utnubu/manifests/oras.unknown":
  unexpected status code 400: unsupported:
  Unsupported manifest type: "application/vnd.oci.artifact.manifest.v1+json"
```

vs

```
% git grep ErrorCodeUnsupported
registry/remote/errcode/errors.go:42:   ErrorCodeUnsupported         = "UNSUPPORTED"
```

Reply:
I think it aligns with the casing here - https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes

Reply:
@nima Can you help confirm if the fallback works for ECR? I notice that the in-body error message was not capitalized in ECR, shown as below:

> unexpected status code 400: unsupported: Unsupported manifest type: "application/vnd.oci.artifact.manifest.v1+json"


You can `make` from source and run a `oras attach` based on ECR